src/Covenant/ASG.hs

```diff
@@ -897,49 +906,93 @@ cata rAlg rVal =
+                      pure $ Datatype bfName . Vector.singleton $ lastTyArg
+                    _ -> throwError . CataWrongBuiltinType $ bT
+                  Datatype tyName tyVars -> do
+                    lookedUp <- asks (view (#datatypeInfo % at tyName))
+                    case lookedUp of
+                      Nothing -> throwError . CataNoSuchType $ tyName
+                      Just info -> case view #baseFunctor info of
+                        Just (DataDeclaration actualBfName _ _ _, _) -> do
+                          unless (bfName == actualBfName) (throwError . CataUnsuitable algT $ valT)
+                          let lastTyArg' = stepDownDB lastTyArg
+                          pure . Datatype bfName . Vector.snoc tyVars $ lastTyArg'
+                        _ -> throwError . CataNoBaseFunctorForType $ tyName
+                  _ -> throwError . CataWrongValT $ valT
+                resultT <- tryApply algT appliedArgT
+                refTo . AValNode resultT . CataInternal rAlg $ rVal
+              _ -> throwError . CataNotAnAlgebra $ t
+          _ -> throwError . CataNonRigidAlgebra $ algT
         t -> throwError . CataNotAnAlgebra $ t
     t -> throwError . CataApplyToNonValT $ t
 
 -- Helpers
 
+-- Note (Koz, 13/08/2025): We need this procedure specifically for `cata`. The
+-- reason for this has to do with how we construct the 'base functor form' of
+-- the value to be torn down by the catamorphism, in order to use the
+-- unification machinery to get the type of the final result.
+--
+-- To be specific, suppose we have `<List_F r (Maybe r) -> !Maybe r>` as our algebra
+-- argument (where `r` is some rigid), and `List r` as the value to be torn
+-- down. If we assume the rigid is bound one scope away, `r`'s DeBruijn index
+-- will be `S Z` for
+-- the value to be torn down, but `S (S Z)` for the algebra argument. The way
+-- our approach works is:
+--
+-- 1. Look at the algebra argument, specifically the base functor type. Take its
+--    last type argument, which we will call `last`.
+-- 2. Determine the base functor for the value to be torn down. Cook up a new
+--    instance of the base functor type, copying all the type arguments from the
+--    value to be torn down in the same order. Then put `last` at the end.
+-- 3. Force the algebra argument thunk, then try and apply the result of Step 2
+--    to that.
+--
+-- However, if `last` is a rigid, we have an 'off by one error'. To see why,
+-- consider the form of the algebra argument:
+--
+-- `ThunkT . Comp0 $ Datatype "List_F" [tyvar (S (S Z)) ix0, ....`
+--
+-- However, `tyvar (S (S Z)) ix0` is not valid in the scope of the value to be
+-- torn down: that same rigid would have DeBruijn index `S Z` there instead.
+-- This applies the same if the tyvar is part of a datatype.
+--
+-- As we prohibit non-rigid algebras, this requires us to lower the DeBruijn
+-- index by one for our process.
+stepDownDB :: ValT AbstractTy -> ValT AbstractTy
```

Comment:
This is for the final type argument of the algebra, right? 

The only thing that gives me some pause here is that I'm not immediately sure why you don't traverse the thing you're stepping down if it's a Thunk. Couldn't a thunk contain abstractions which are subject to this off-by-one error? (I'm honestly not sure.)